bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java

```diff
@@ -583,6 +583,49 @@ public static Pipeline initializeForWrite(Pipeline p) {
     return p;
   }
 
+  private static abstract class AbstractCloudBigtableTableWriteFn<In, Out> extends DoFn<In, Out> {
+    private static final long serialVersionUID = 1L;
+    protected final Logger LOG = LoggerFactory.getLogger(getClass());
+    protected CloudBigtableConfiguration config;
+    protected transient Connection conn;
+
+    public AbstractCloudBigtableTableWriteFn(CloudBigtableConfiguration config) {
+      this.config = config;
+    }
+
+    /**
+     * Creates a {@link Connection} and a {@link BufferedMutator}.
+     */
+    @Override
+    public void startBundle(Context context) throws Exception {
+      LOG.debug("Creating the Bigtable connection.");
+      conn = new BigtableConnection(config.toHBaseConfig());
+    }
+
+    @Override
+    public void finishBundle(DoFn<In, Out>.Context c) throws Exception {
+      LOG.debug("Closing the Bigtable connection.");
+      conn.close();
+    }
+
+    /**
+     * Logs the {@link Context} and the exception's
+     * {@link RetriesExhaustedWithDetailsException#getExhaustiveDescription()}.
+     */
+    protected void logExceptions(Context context, RetriesExhaustedWithDetailsException exception) {
```

Comment:
The following two methods don't belong in this class, no? in a util class perhaps?